torch/_functorch/aot_autograd.py

```diff
@@ -3421,20 +3421,28 @@ def convert(idx, x):
 
         fake_flat_args = process_inputs(flat_args)
 
-        needs_autograd = (
-            any(x.requires_grad for x in fake_flat_args if isinstance(x, Tensor))
-            and torch.is_grad_enabled()
-        )
-
         with enable_python_dispatcher():
             # Patch set_rng_state as set_rng_state with fake tensors is
             # nonsensical. This does not affect the collection of metadata.
             with patch("torch.cuda.set_rng_state", lambda *args: None):
                 fw_metadata = run_functionalized_fw_and_collect_metadata(
                     flat_fn,
-                    keep_input_mutations=aot_config.keep_inference_input_mutations and not needs_autograd,
+                    keep_input_mutations=False # delay this since we want to set it based off of needs_autograd
```

Comment:
Does this mean you have to reanalyze this if it turns out it doesn't need autograd?

Reply:
No, `run_functionalized_fw_and_collect_metadata` does not use `keep_input_mutations` at all, except for constructing the metadata object.

Reply:
yeah we will need to re-run the analyze.


Although sigh, this is my fault. For AOTDispatch I'm currently relying on this info to decide how many fw outs to compute subclass information for (since training graphs have more outputs than inference graphs). Some time after PTC I'll refactor `run_functionalized_fw_and_collect_metadata` to not have to do this.

Reply:
It does as of this morning (this is the hack that I landed to get DTensor support earlier that I want to clean up soon): https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1245